client/src/controllers/SearchController.ts

```diff
@@ -0,0 +1,185 @@
+            `Cannot find results element at "${this.resultsValue}"`,
+        ]
+          .filter(Boolean)
+          .join(', '),
+      );
+    }
+
+    this.resultsElement = resultsElement as HTMLElement;
+
+    // set up icons
+    const iconContainer = (
+      this.hasInputTarget ? this.inputTarget : this.element
+    ).parentElement;
+
+    this.icon = iconContainer?.querySelector('use') || null;
+    this.iconName = this.icon?.getAttribute('href') || null;
+    this.loadingValue = false;
+
+    // set up debounced search action method
+    this.search = debounce(this.runSearch.bind(this), this.waitValue);
+  }
+
+  disconnect() {
+    this.loadingValue = false;
+    this.search.cancel();
+  }
+
+  /**
+   * Abort any existing requests & set up new abort controller, then fetch and patch
+   * in the HTML results, dispatching an event and handling any clean up.
+   * Cache the current search index so we can ensure that late responses do not
+   * overtake earlier results.
+   */
+  async fetchResults(param: string | CustomEvent<{ url: string }>) {
+    if (this.abortController) this.abortController.abort();
+    this.abortController = new AbortController();
+    const { signal } = this.abortController;
+
+    /** Parse a request URL from the supplied param, as a string or inside a custom event */
+    const requestUrl =
+      typeof param === 'string' ? param : param.detail.url || '';
+
+    this.loadingValue = true;
+
+    return fetch(requestUrl, {
+      headers: { 'x-requested-with': 'XMLHttpRequest' },
+      signal,
+    })
+      .then((response) => {
+        if (!response.ok) {
+          throw new Error(`HTTP error! Status: ${response.status}`);
+        }
+        return response.text();
+      })
+      .then((results) => {
+        this.resultsElement.innerHTML = results;
+
+        this.dispatch('success', {
+          cancelable: false,
+          detail: { results },
+          // Stimulus dispatch target element type issue https://github.com/hotwired/stimulus/issues/642
+          target: this.resultsElement as HTMLInputElement,
+        });
+      })
+      .catch((error) => {
+        if (error.name === 'AbortError') return;
+        // eslint-disable-next-line no-console
+        console.error(`Error fetching ${requestUrl}`, error);
+        delete this.abortController;
+      })
+      .finally(() => {
+        this.loadingValue = false;
+      });
+  }
+
+  loadingValueChanged(isLoading: boolean) {
+    if (!this.icon) return;
+
+    if (isLoading) {
+      this.icon.setAttribute('href', '#icon-spinner');
+    } else if (this.iconName) {
+      this.icon.setAttribute('href', this.iconName);
+    }
+  }
+
+  /**
+   * Perform a search based on a single input query, and only if that query's value
+   * differs from the current matching URL param. Once complete, update the URL param.
+   * Additionally, clear the 'page' param in the URL if present.
+   */
+  runSearch() {
+    const pageParam = this.pageParamValue;
```

Comment:
Instead of being tied to the input, it probably makes more sense for this to serialise the form, which should contain a `p` reset, merge this with the url values (so we don't overwrite existing ones) and then apply the request based on the form. 

This means p may get requested also. Might be ok.